test/integration/resources/init_timers.c

```diff
@@ -0,0 +1,173 @@
+#define TIMER_BASE_ADDRESS 0x10U
+#define TIMER_SIZE 0x10U
+#define TIMEOUT 100000
+#define TIMER1_SCALER_RELOAD_VALUE 0x1FFU
+#define TIMER2_SCALER_RELOAD_VALUE 0xFFU
+#define TIMER_COUNTER_UNDERFLOW_VALUE -1U
+#define TIMER_COUNTER_START_VALUE 0xFFU
+#define TIMER_ENABLE 0x01
+
+#define TIMER1_SIZE 4
+#define TIMER2_SIZE 2
+
+typedef volatile struct
+{
+  uint32_t scaler;
+  uint32_t reloadScaler;
+  uint32_t configuration;
+} *CoreUnit;
+
+typedef volatile struct
+{
+  uint32_t timerCounter;
+  uint32_t timerReloadCounter;
+  uint32_t timerControl;
+} *TimerUnit;
+
+typedef volatile struct
+{
+  CoreUnit core;
+  TimerUnit timers[TIMER1_SIZE];
+} Timer1Register;
+
+typedef volatile struct
+{
+  CoreUnit core;
+  TimerUnit timers[TIMER2_SIZE];
+} Timer2Register;
+
+Timer1Register timer1;
+Timer2Register timer2;
+
+void sendMsg (const char *msg)
+{
+  for (size_t i = 0; i < strlen(msg); i++) {
+    rtems_putc(msg[i]);
+  }
+}
+
+void checkTimersWorking (volatile TimerUnit *timers, int timersSize, int *timersUnderflowedArray, int timestamp)
+{
+  if (timersUnderflowedArray[1] == 0){
+    timersUnderflowedArray[0] = 0;
+    for (int i = 0; i < timersSize; i++) {
+      if (timers[i]->timerCounter == TIMER_COUNTER_UNDERFLOW_VALUE) {
+        timersUnderflowedArray[0]++;
+      }
+    }
+
+    if (timersUnderflowedArray[0] == timersSize) {
+      timersUnderflowedArray[1] = timestamp;
+    }
+  }
+}
+
+static void timerInit ()
+{
+  timer1.core = (CoreUnit) TIMER1_CORE_ADDRESS;
+  timer1.core->reloadScaler = TIMER1_SCALER_RELOAD_VALUE;
+  for (size_t i = 0; i < TIMER1_SIZE; i++) {
+    timer1.timers[i] = (TimerUnit) (TIMER1_CORE_ADDRESS + TIMER_BASE_ADDRESS + (i * TIMER_SIZE));
+    timer1.timers[i]->timerCounter = TIMER_COUNTER_START_VALUE;
+    timer1.timers[i]->timerControl = TIMER_ENABLE;
+  }
+
+  timer2.core = (CoreUnit) TIMER2_CORE_ADDRESS;
+  timer2.core->reloadScaler = TIMER2_SCALER_RELOAD_VALUE;
+  for (size_t i = 0; i < TIMER2_SIZE; i++) {
+    timer2.timers[i] = (TimerUnit) (TIMER2_CORE_ADDRESS + TIMER_BASE_ADDRESS + (i * TIMER_SIZE));
+    timer2.timers[i]->timerCounter = TIMER_COUNTER_START_VALUE;
+    timer2.timers[i]->timerControl = TIMER_ENABLE;
+  }
+}
+
+static void Init( rtems_task_argument arg )
+{
+  (void) arg;
+
+  timerInit ();
+
+  int timers1Underflowed[2] = {0, 0};
+  int timers2Underflowed[2] = {0, 0};
+  int timeout = 0;
+  while (((timers1Underflowed[0] != TIMER1_SIZE) || (timers2Underflowed[0] != TIMER2_SIZE)) && (timeout != TIMEOUT)) {
+    checkTimersWorking (timer1.timers, TIMER1_SIZE, timers1Underflowed, timeout);
+    checkTimersWorking (timer2.timers, TIMER2_SIZE, timers2Underflowed, timeout);
+    timeout++;
+  }
+
+  bool timersEndsInProperOrder = false;
+  float scalerDifferenceFactor = (float) TIMER1_SCALER_RELOAD_VALUE / (float) TIMER2_SCALER_RELOAD_VALUE;
+  float differenceFactor = (float) timers1Underflowed[1] / (float) timers2Underflowed[1];
+
+  if (differenceFactor > scalerDifferenceFactor) {
```

Comment:
Should this test be >, or rather (abs(diff) < epsilon)?

Reply:
ack

Reply:
OK, maybe the pseudocode was not clear.
const float desiredScaleFactor = (float) TIMER1_SCALER_RELOAD_VALUE / (float) TIMER2_SCALER_RELOAD_VALUE;
const float actualScaleFactor = (float) timer1Status.expirationTime / (float) timer2Status.expirationTime;
const float scaleDifference = abs(actualScaleFactor  - desiredScaleFactor );
if (scaleDifference < 0.0001f) // TODO: establish practical accuracy

Reply:
ack